Guard HIR lowered contracts with `contract_checks`

dawidl022 opened the pull request:
Refactor contract HIR lowering to ensure no contract code is executed when contract-checks are disabled.

The call to `contract_checks` is moved to inside the lowered fn body, and contract closures are built conditionally, ensuring no side-effects present in contracts occur when those are disabled. This partially addresses rust-lang/rust#139548, i.e. the bad behavior no longer happens with contract checks disabled (`-Zcontract-checks=no`).

The change is made in preparation for adding contract variable declarations - variables declared before the `requires` assertion, and accessible from both `requires` and `ensures`, but not in the function body (PR rust-lang/rust#144444). As those declarations may also have side-effects, it's good to guard them with `contract_checks` - the new lowering approach allows for this to be done easily.

Contracts tracking issue: rust-lang/rust#128044

**Known limiatations**:

- It is still possible to early return from the *function* from within a contract, e.g.

  ```rust
  #[ensures({if x > 0 { return 0 }; |_| true})]
  fn foo(x: u32) -> i32 {
      42
  }
  ```

  When `foo` is called with an argument greater than 0, instead of `42`, `0` will be returned.

  As this is not a regression, it is not addressed in this PR. However, it may be worth revisiting later down the line, as users may expect a form of early return from *contract specifications*, and so returning from the entire *function* could cause confusion.

- ~Contracts are still not optimised out when disabled. Currently, even when contracts are disabled, the code generated causes existing optimisations to fail, meaning even disabled contracts could impact runtime performance. This issue is blocking rust-lang/rust#136578, and has not been addressed in this PR, i.e. the `mir-opt` and `codegen` tests that fail in rust-lang/rust#136578 still fail with these new HIR lowering changes.~ Contracts should now be optimised out when disabled, however some regressions tests still need to be added to be sure that is indeed the case.

dawidl022 committed:
Guard HIR lowered contracts with contract_checks

Refactor contract HIR lowering to ensure no contract code is
executed when contract-checks are disabled.

The call to contract_checks is moved to inside the lowered fn
body, and contract closures are built conditionally, ensuring
no side-effects present in contracts occur when those are disabled.

celinval commented:
What if we remove the `contract_checks` for now? I think this will be much simpler and it won't add extra optimization overhead which may even fix the issue from #136578.

dawidl022 commented:
> What if we remove the `contract_checks` intrinsic for now? Instead, we only lower the contracts if the user passes `-Zcontract-checks=yes`. I think this will be much simpler and it won't add extra optimization overhead it will likely fix the issue from #136578.

I think this would indeed fix the issue. I wanted to try it out, but wasn't sure how to access compiler flags (e.g. `-Zcontract-checks=yes`) from the lowering code (or the parser for that matter). Any advice on how to achieve this?

celinval commented:
You should be able to access them through `self.tcx.sess.opts.unstable_opts`.

RalfJung commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -2569,9 +2569,10 @@ pub const unsafe fn const_make_global(ptr: *mut u8) -> *const u8 {
 #[unstable(feature = "contracts_internals", issue = "128044" /* compiler-team#759 */)]
+#[rustc_const_unstable(feature = "contracts", issue = "128044")]
 #[inline(always)]
+#[lang = "contract_checks"]
 #[rustc_intrinsic]
```
Making the same thing both a lang item and an intrinsic doesn't make a lot of sense. Why do you propose to do this?

dawidl022 replied:
> Making the same thing both a lang item and an intrinsic doesn't make a lot of sense.

Could you please elaborate why this doesn't make sense, or point me to sources where I can read up on this? I don't really understand the relationship between the two.

> Why do you propose to do this?

I added the lang item annotation so that I can generate HIR code referring to `contract_checks`, and did not think there was anything wrong with that, given that both `contract_check_requires` and `contract_check_ensures` are already marked both as lang items and intrinsics.

dawidl022 commented:
> What if we remove the contract_checks intrinsic for now? Instead, we only lower the contracts if the user passes -Zcontract-checks=yes. I think this will be much simpler and it won't add extra optimization overhead it will likely fix the issue from https://github.com/rust-lang/rust/pull/136578.

I'm happy to implement it that way, and if there is consensus, close this this PR in favour of that solution. I imagine that even with `-Zcontract-checks=no`, we'd still want to typecheck the contracts (at least in the debug build) to ensure they are well-formed. I'm wondering what would be the best way to achieve this, given typechecking occurs after HIR lowering. 

Maybe the "typechecking when disabled issue"  is not a priority right now, given that we're still figuring out what exactly the contract language should be (i.e. what extra rules contracts have to obey w.r.t. regular rust code). If that's the case, I'm happy to proceed with whatever makes most sense for now.

Right now, when naively guarding the call to `lower_contract` with `self.tcx.sess.opts.unstable_opts`  I'm getting a regression fixed by #136837, so I need to figure out what exactly is going on there with associated items.

celinval commented on compiler/rustc_ast_lowering/src/contract.rs:
```diff
@@ -0,0 +1,260 @@
+                // into:
+                //
+                // {
+                //      let __postcond = if contracts_checks() {
```
The suggestion is for this code to be either:
```rust
let __postcond = if false {
  // ...
}
```
or 
```rust
let __postcond = if true {
  // ...
}
```
depending on the value of the `-Z contract-checks`

celinval replied:
The reason for this to be an intrinsic is for tools / custom backends to be able to override the behavior

celinval replied:
Why would you still need the lang item?

dawidl022 committed:
Guard HIR contracts based on compiler flag rather than lang_item

This allows the optimiser to properly eliminate contract code
when runtime contract checks are disabled.

It comes at the cost of having to recompile upstream crates
(e.g. std) to enable contracts in them. However, this trade
off is acceptable if it means disabled runtime contract checks
do not affect the runtime performance of the functions they annotate.

With the proper elimination of contract code, which this change
introduces, the runtime performance of annotated functions
should be the same as the original unannotated function.

dawidl022 committed:
Remove no longer used contract_checks intrinsic

The contract_checks compiler flag is now used to determine
if runtime contract checks should be enabled, as opposed
to the compiler intrinsic as previously.

dawidl022 committed:
Wrap contract clauses in brances instead of parenthesis

The compiler complained about uncecessary parenthesis on contract clauses,
which were insterted by the contract macros. This commit changes the
macro to use braces as the delimiter instead, fixing the issue.

dawidl022 commented:
I went for:
> Lower body once, but wrap call to contract_check_ensures

and the contracts in https://github.com/rust-lang/rust/pull/136578 now pass the `codegen-llvm` tests.

What would be a good way of rigorously testing whether contracts are indeed fully optimised out with this change, as opposed to relying on the observation that one optimisation test that was failing before now succeeds?

celinval commented:
Hey @dawidl022, yes, we should remove the symbol and type check

dawidl022 commented:
Thanks for clarifying @celinval. I've removed `sym::contract_checks` and any code that depended on it.

> What would be a good way of rigorously testing whether contracts are indeed fully optimised out with this change, as opposed to relying on the observation that one optimisation test that was failing before now succeeds?

Do you have any thoughts on this?

celinval commented:
That's a good question. I would say you first need to check if the contract is pruned by the MIR optimizations or by LLVM optimizations. If MIR optimizations are able to prune the contracts, I would create a [mir-opt test](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html?highlight=assembly-llvm#mir-opt-tests). If the code is optimized by LLVM, you can create an [assembly test](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html?highlight=assembly-llvm#assembly-tests) like [this one](https://github.com/rust-lang/rust/blob/master/tests/assembly-llvm/is_aligned.rs).

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17824373651/job/50674028546) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50674028546) [(plain)](https://github.com/rust-lang/rust/commit/c0af37682caa1f2219e742ed866f7c958abfcdae/checks/50674028546/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: ui test did not emit an error
note: by default, ui tests are expected not to compile.
hint: use check-pass, build-pass, or run-pass directive to change this behavior.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/feature-gates/feature-gate-cfg-contract-checks.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-cfg-contract-checks" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
warning: unexpected `cfg` condition name: `contract_checks`
##[warning]  --> /checkout/tests/ui/feature-gates/feature-gate-cfg-contract-checks.rs:4:10
   |
LL |     cfg!(contract_checks) //~ ERROR `cfg(contract_checks)` is experimental
   |          ^^^^^^^^^^^^^^^
   |
   = help: expected names are: `FALSE` and `test` and 30 more
   = help: to expect this configuration use `--check-cfg=cfg(contract_checks)`
   = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg.html> for more information about checking conditional configuration
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #146683) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

celinval commented:
> > What if we remove the contract_checks intrinsic for now?
> 
> @celinval what clean up needs to be done for this? I removed the intrinsic from `intrinsics/mod.rs`, but `contract_checks` also appears in `sym`, and its signature is present in `check_intrinsic_type`. Do we remove the `sym` as well?

Shoot, I just realized that this removed the configuration `#[cfg(contract_checks)]`. Sorry about that. We still want users to be able to use the config.

bors commented:
:umbrella: The latest upstream changes (presumably #146999) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented on library/core/src/intrinsics/mod.rs:
```diff
@@ -2689,16 +2672,24 @@ pub const fn contract_check_requires<C: Fn() -> bool + Copy>(cond: C) {
 #[lang = "contract_check_ensures"]
 #[rustc_intrinsic]
-pub const fn contract_check_ensures<C: Fn(&Ret) -> bool + Copy, Ret>(cond: C, ret: Ret) -> Ret {
+pub const fn contract_check_ensures<C: Fn(&Ret) -> bool + Copy, Ret>(
```
Update the docs on this explaining that None for cond means no post condition checking

oli-obk reviewed (approved):
lgtm

One thing I was wondering about was whether you think a test with an aux crate would be useful for testing the case that a dep has different compile time flags from the main crate

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/144438/5b9007bfc358817cf066ee27c6b440440727d3a7..085ca0e122467255f9281cb0c69948cfd21189d9)*

dawidl022 committed:
Test cross-crate runtime contract checks

These tests capture the behaviour that the decision to include/exclude
runtime contract assertions is determined on a per-crate basis,
i.e. by the flags used to compile each crate.

oli-obk commented on tests/ui/contracts/cross-crate-checks/auxiliary/id.rs:
```diff
@@ -0,0 +1,16 @@
+//@ [unchk_pass] compile-flags: -Zcontract-checks=no
+//@ [unchk_fail] compile-flags: -Zcontract-checks=yes
+//@ [chk_pass] compile-flags: -Zcontract-checks=no
+//@ [chk_fail] compile-flags: -Zcontract-checks=yes
```
I didn't realize it was possible to make aux tests also handle revisions from the aux-test user. Considering the main test properly crashes or passes in all the cases configured here, it does seem to work flawlessly
